### PR TITLE
Decommissioning Change Plan

### PR DESCRIPTION
**Pull request by clairecfu**

### CODE REVIEW

File: rackcity/models/asset.py [hunk 1]
```diff
@@ -55,7 +55,8 @@ def validate_hostname_uniqueness(value, asset_id, change_plan, related_asset):
     assets, assets_cp = get_assets_for_cp(change_plan.id)
     matching_assets = assets_cp.filter(hostname=value, change_plan=change_plan)
     if (
-        len(matching_assets) > 0
+        value
+        and len(matching_assets) > 0
```
Comment (clairecfu) [inline rackcity/models/asset.py:59]: CRAZY hostname validation bug for assetCPs -> was flagging empty hostnames as non unique

File: rackcity/views/change_plan_views.py [hunk 2]
```diff
@@ -470,20 +470,23 @@ def change_plan_execute(request, id):
...
+            if asset_cp.is_decommissioned:
+                failure_response = decommission_asset_cp(
+                    updated_asset, asset_cp, change_plan,
+                )
+                if failure_response:
+                    return failure_response
+                num_decommissioned += 1
+                log_action(
+                    request.user, None, Action.DECOMMISSION, change_plan=change_plan,
+                )
```
Comment (clairecfu) [inline rackcity/views/change_plan_views.py:489]: these are only changes i made with regard to decommissioning, all else is bens

File: rackcity/models/asset.py (hunk 1 above)
Comment (juliasaveliff) [reply rackcity/models/asset.py:59]: OH DEAR

File: rackcity/views/change_plan_views.py [hunk 3]
```diff
@@ -470,20 +470,23 @@ def change_plan_execute(request, id):
...
-            if failure_response:
-                return failure_response
-            num_decommissioned += 1
-            log_action(
-                request.user, None, Action.DECOMMISSION, change_plan=change_plan,
-            )
+    ## decomission blades first
+    blades_cp = assets_cp.filter(model__model_type=ModelType.BLADE_ASSET.value)
+    rackmount_cp = assets_cp.filter(~Q(model__model_type=ModelType.BLADE_ASSET.value))
+    for asset_cp_query in [blades_cp, rackmount_cp]:
```
Comment (juliasaveliff) [inline rackcity/views/change_plan_views.py:476]: I'm assuming you tested that this worked, right? I've run into some weird behavior with concatenating query sets into a list before

Comment (juliasaveliff) [reply rackcity/views/change_plan_views.py:476]: never mind lmao i missed the line right below this one

File: rackcity/views/decommission_views.py [hunk 4]
```diff
@@ -118,137 +129,160 @@ def decommission_asset_parameterized(data, query_params, user):
...
             blades = assets.filter(chassis=asset_id)
         else:
             blades = assets_cp.filter(chassis=asset_id)
-
         for blade in blades:
-            data_for_blade = data.copy()
-            data_for_blade["id"] = blade.id
             try:
-                decommission_asset_parameterized(data_for_blade, query_params, user)
+                decommission_asset_parameterized(
```
Comment (juliasaveliff) [inline rackcity/views/decommission_views.py:162]: terrifying recursion, innit!

**Pull request merged**